ksp_plugin/geometric_potential_plotter.cpp

```diff
@@ -0,0 +1,213 @@
+using namespace principia::geometry::_space;
+using namespace principia::integrators::_embedded_explicit_runge_kutta_integrator;  // NOLINT
+using namespace principia::integrators::_methods;
+using namespace principia::numerics::_global_optimization;
+using namespace principia::numerics::_root_finders;
+using namespace principia::physics::_rotating_pulsating_reference_frame;
+using namespace principia::quantities::_named_quantities;
+using namespace principia::quantities::_quantities;
+using namespace principia::quantities::_si;
+
+GeometricPotentialPlotter::GeometricPotentialPlotter(
+    not_null<Ephemeris<Barycentric>*> ephemeris)
+    : ephemeris_(ephemeris) {}
+
+void GeometricPotentialPlotter::Interrupt() {
+  plotter_ = jthread();
+  // We are single-threaded here, no need to lock.
+  plotter_idle_ = true;
+}
+
+void GeometricPotentialPlotter::RequestEquipotentials(
+    Parameters const& parameters) {
+  last_parameters_ = parameters;
+  absl::MutexLock l(&lock_);
+  // Only process this request if there is no analysis in progress.
+  if (plotter_idle_) {
+    plotter_idle_ = false;
+    plotter_ = MakeStoppableThread(
+        [this, parameters]() { PlotEquipotentials(parameters).IgnoreError(); });
+  }
+}
+
+std::optional<GeometricPotentialPlotter::Parameters> const&
+GeometricPotentialPlotter::last_parameters()
+    const {
+  return last_parameters_;
+}
+
+void GeometricPotentialPlotter::RefreshEquipotentials() {
+  absl::MutexLock l(&lock_);
+  if (next_equipotentials_.has_value()) {
+    equipotentials_ = std::move(next_equipotentials_);
+    next_equipotentials_.reset();
+  }
+}
+
+GeometricPotentialPlotter::Equipotentials const*
+GeometricPotentialPlotter::equipotentials() const {
+  return equipotentials_.has_value() ? &*equipotentials_ : nullptr;
+}
+
+absl::Status GeometricPotentialPlotter::PlotEquipotentials(
+    Parameters const& parameters) {
+  Instant const& t = parameters.time;
+
+  auto const& primary =
+      parameters.primary->mass() > parameters.secondary->mass()
+          ? *parameters.primary
+          : *parameters.secondary;
+  auto const& secondary =
+      parameters.primary->mass() > parameters.secondary->mass()
+          ? *parameters.secondary
+          : *parameters.primary;
+
+  RotatingPulsatingReferenceFrame<Barycentric, Navigation> const
+      reference_frame(ephemeris_, parameters.primary, parameters.secondary);
+  auto const plane =
+      Plane<Navigation>::OrthogonalTo(Vector<double, Navigation>({0, 0, 1}));
+
+  auto const potential = [&reference_frame,
+                          &t](Position<Navigation> const& position) {
+    return reference_frame.GeometricPotential(t, position);
+  };
+  auto const acceleration = [&reference_frame,
+                             &t](Position<Navigation> const& position) {
+    auto const acceleration = reference_frame.GeometricAcceleration(
+        t, {position, Velocity<Navigation>{}});
+    // Note the sign.
```

Comment:
I noted it, but maybe explain why it's there.

Reply:
Done, and renamed the λ.